### PR TITLE
Assumereadtype EquiJoin branches

### PR DESCRIPTION
**Pull request by rintcius**
Part of #3048

### CODE REVIEW

File: mongodb/src/main/scala/quasar/physical/mongodb/MongoDbPlanner.scala [hunk 1]
```diff
@@ -1230,16 +1230,22 @@ object MongoDbPlanner {
...
     for {
-      mongoQS0 <- Trans(assumeReadType[T, MQS, M](Type.AnyObject), qs)
-      mongoQS1 <- mongoQS0.transCataM(elideQuasarSigil[T, MQS, M](anyDoc))
-      mongoQS2 =  mongoQS1.transCata[T[MQS]](R.normalizeEJ[MQS])
-      mongoQS3 =  BR.branches.modify(
-        _.transCata[FreeQS[T]](liftCo(R.normalizeEJCoEnv[QScriptTotal[T, ?]]))
-        )(mongoQS2.project).embed
+      mongoQS1 <- Trans(assumeReadType[T, MQS, M](Type.AnyObject), qs)
+      mongoQS2 <- mongoQS1.transCataM(elideQuasarSigil[T, MQS, M](anyDoc))
+      mongoQS3 <- fixpointM(normalize, mongoQS2)
```
Comment (rintcius) [inline mongodb/src/main/scala/quasar/physical/mongodb/MongoDbPlanner.scala]: This is getting trickier than I like it to be. Some constraints:

* `elideQuasarSigil` should only be applied once
* `elideQuasarSigil` needs `assumeReadType` to be applied in order to work properly in all cases
* applying `normalize` once does not simplify as much as running until fixpoint does

I kept the phases as-is in this PR, but a better setup may involve changing the phases.

**Comment by alissapajer** [comment]
Needs version label.

**Comment by jsantos17** [comment]
Just a heads up, I temporarily changed the type [here](https://github.com/quasar-analytics/quasar/blob/master/frontend/src/main/scala/quasar/frontend/logicalplan/LogicalPlanR.scala#L309) from `Type.Top` to `Type.AnyObject` and we might still have non-elided `Guard`s in a few places:

1. plan 3-way equi-join ($lookup)
2. plan 3-way equi-join (map-reduce)
3. plan simple inner equi-join with expression ($lookup)
4. plan simple inner equi-join with pre-filtering ($lookup)

**Comment by rintcius** [comment]
@jsantos17 Yes, `assumeReadType` is still not complete. This is due to the difference in coverage between these 2 functions:

```
  def isRewrite[T[_[_]]: BirecursiveT: EqualT, F[_]: Functor, G[_]: Functor, A](GtoF: PrismNT[G, F], qs: G[A])(implicit
    QC: QScriptCore[T, ?] :<: F,
    SR: Const[ShiftedRead[AFile], ?] :<: F,
    ev: Recursive.Aux[A, G])
      : Boolean = qs match {
    case GtoF(QC(Filter(Embed(GtoF(SR(Const(ShiftedRead(_, ExcludeId))))), _)))
       | GtoF(QC(Sort(Embed(GtoF(SR(Const(ShiftedRead(_, ExcludeId))))), _, _)))
       | GtoF(QC(Sort(Embed(GtoF(QC(Filter(Embed(GtoF(SR(Const(ShiftedRead(_, ExcludeId))))), _)))), _ , _)))
       | GtoF(QC(Subset(_, FreeQS(SRT(ShiftedRead(_, ExcludeId))), _ , _)))
       | GtoF(SR(Const(ShiftedRead(_, ExcludeId)))) => true
    case _ => false
  }

  def isRewriteFree[T[_[_]]](fq: FreeQS[T]): Boolean = fq match {
    case FreeQS(SRT(ShiftedRead(_, ExcludeId))) => true
    case _ => false
  }
```

`isRewriteFree` only recognizes a simple `ShiftedRead` as shape preserving whereas `isRewrite` recognizes more patterns. It would be great if we'd be able to recognize the same patterns in `isRewriteFree`. BTW, _maybe_ also the `TODO` I mentioned in `UnaryFunctions` would help out here.

File: connector/src/main/scala/quasar/qscript/BranchesMap.scala [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+import matryoshka._
+import matryoshka.patterns._
+
+import monocle._
+import scalaz._, Scalaz._
+
+// TODO look into traversing FreeQS as well to simplify rewrites like
+// assumeReadType
+trait BranchesMap[T[_[_]], IN[_]] {
```
[the remark this answers is not in the document]
Comment (rintcius) [reply connector/src/main/scala/quasar/qscript/BranchesMap.scala]: As discussed renamed to `UnaryFunctions`

File: foundation/src/main/scala/quasar/contrib/scalaz/package.scala [hunk 3]
```diff
@@ -40,6 +40,17 @@ package object scalaz {
     }
   }
 
+  def nextM[M[_]: Monad, A, B, C](f: (A, B) => C, m: A => M[B], a: A): M[(B, C)] =
+    m(a).map(b => (b, f(a, b)))
+
+  @SuppressWarnings(Array("org.wartremover.warts.Recursion"))
+  def fixpointM[M[_]: Monad, A: Equal](f: A => M[A], a: A): M[A] =
```
[the remark this answers is not in the document]
Comment (rintcius) [reply foundation/src/main/scala/quasar/contrib/scalaz/package.scala]: Right, I wasn't really happy with this place and was wondering about impl too. I was able to remove the need for it though, so just removed the code again

**Comment by rintcius** [comment]
@wemrysi I moved `assumeReadType` to the end in `normalize`. Looks like in this way we can avoid having to run until fixpoint. I have added an example when `assumeReadType` needs to be run after `R.normalizeEJ/R.normalizeEJCoEnv`

**Comment by jsantos17** [comment]
@rintcius just thinking out loud here, but `isRewrite` looks like a good candidate for refactoring into a prism that focuses on `ShiftedRead`s nested inside shape preserving transformation. It would have a signature somewhat like `FreeQS[T] => Option[Const[ShiftedRead]]` with the reverse of the prism simply injecting `Const[ShiftedRead]` into `QScriptTotal` and then lifting into `Free`.

Not sure if this is law abiding tho, but it does seem correct intuitively.

**Pull request merged**